docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+
+1. **Interface**: In the context of this document, "interface" refers to Go's interface concept.
+
+2. **Module**: In this document, "module" refers to a Go module.
+
+3. **Package**: In the context of Go, a "package" refers to a unit of code organization.
+
+## Context
+
+In order to fully understand the need for changes and the proposed improvements, it's crucial to consider the current state of affairs:
+
+* The Cosmos SDK currently lacks a comprehensive ADR for the cryptographic package.
+* Type leakage of specific crypto data types expose backward compatibility and extensibility challenges.
+* The demand for a more flexible and extensible approach to cryptography and address management is high.
+* Architectural changes are necessary to resolve many of the currently open issues.
+* There is a current trend towards modularity in the Interchain stack (e.g. runtime modules)
+* Security implications are a critical consideration during the redesign work.
+
+## Objectives
+
+The key objectives for the CryptoV2 module are:
+
+Modular Design Philosophy
+
+* Establish a flexible and extensible foundation using interfaces to enable the seamless integration of various cryptographic methods and tools.
+
+* Restructure, Refactor, and Decouple: Update the cryptography codebase to ensure modularity and future adaptability.
+
+Documentation & Community Engagement
+
+* Cryptography v2 ADR: Draft a new Architecture Decision Record to guide and document the evolution of the module (this document).
+
+* Enhance documentation to ensure clarity, establish a good practices protocol and promote community engagement, providing a platform for feedback and collaborative growth.
+
+Backward Compatibility & Migration
+
+* Prioritize compatibility with previous module version to avoid disruptions for existing users.
+
+* Design and propose a suitable migration path, ensuring transitions are as seamless as possible.
+
+* Evaluate and decide on the relevance of existing systems and tools, incorporating or deprecating them based on their alignment with the module's new vision.
+
+Developer-Centric Approach
+
+* Prioritize clear, intuitive interfaces and best-practice design principles.
+* Improve Developer Experience: Provide tools, samples, and best practices to foster an efficient and user-friendly development environment.
+
+Leverage Extensibility
+
+* Utilize the module's modular design to support a wide range of cryptographic tools, key types, and methods, ensuring adaptability for future technological advancements.
+* Integrate support for advanced cryptographic features, ensuring the module's position at the forefront of cryptographic technologies.
+
+Quality Assurance
+
+* Enhanced Test Coverage: Improve testing methodologies to ensure the robustness and reliability of the module.
+* Conduct an Audit: After implementation, perform a comprehensive audit to identify potential vulnerabilities and ensure the module's security and stability.
+
+## Technical Goals
```

Comment:
Is there anywhere I can read more about the limits and issues with the current API that motivates V2? These are all fairly high-level goals.

Reply:
Sure, you can give a read to the [crypto package](https://github.com/cosmos/cosmos-sdk/tree/main/crypto) in the sdk code. At its current state adding new curves is virtually impossible without having to do hacks and workarounds. This and also having support for different HW wallets (or any kind of crypto provider 😉 ) in a modular way is also not possible today.